[GPU Process][Filters] Top level SVGFilter should own its FilterResults

shallawa opened the pull request:
#### b860110f0ae42268a9e9aac5975dd6881c866fc7
<pre>
[GPU Process][Filters] Top level SVGFilter should own its FilterResults
<a href="https://bugs.webkit.org/show_bug.cgi?id=256535">https://bugs.webkit.org/show_bug.cgi?id=256535</a>
rdar://109107572

Reviewed by Simon Fraser.

This allows caching the FilterResults along with the SVGFilter in RemoteResourceCache.

SVGFilter::ensureResults() returns FilterResults&amp; but it takes FilterResultsCreator.
FilterResultsCreator returns a std::unique_ptr&lt;FilterResults&gt; which SVGFilter
will maintain.

* Source/WebCore/platform/graphics/filters/FilterResults.h:
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
(WebCore::RenderSVGResourceFilter::markFilterForRepaint):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.h:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::ensureResults):
(WebCore::SVGFilter::clearEffectResult):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:

Canonical link: <a href="https://commits.webkit.org/263996@main">https://commits.webkit.org/263996@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/631ef5b823851aadfb8334cf421a2bb43b786c04

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6295 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6491 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6670 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7869 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6624 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6298 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6711 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6440 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9517 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6407 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6415 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5701 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7939 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3887 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5680 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13568 "25 flakes 129 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5751 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5763 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8010 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6245 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5109 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5657 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1501 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9811 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6026 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilter.h:
```diff
@@ -49,6 +50,10 @@ class SVGFilter final : public Filter {
     FilterEffectVector effectsOfType(FilterFunction::Type) const final;
 
+    FilterResults* results() { return m_results.get(); }
+    WEBCORE_EXPORT FilterResults& ensureResults(const FilterResultsCreator& = [] { return makeUnique<FilterResults>(); });
```
The implementation of this function is this:
```
FilterResults& SVGFilter::ensureResults(const FilterResultsCreator& resultsCreator)
{
    if (!m_results)
        m_results = resultsCreator();
    return *m_results;
}
```
If the default FilterResultsCreator is passed, this is going to be resolved to:
```
FilterResults& SVGFilter::ensureResults(const FilterResultsCreator& resultsCreator)
{
    if (!m_results)
        m_results = makeUnique<FilterResults>();
    return *m_results;
}
```
So this function should compile and it should return a reference to member `*m_results`.

shallawa commented on Source/WebCore/platform/graphics/GraphicsContext.cpp:
```diff
@@ -359,12 +359,13 @@ void GraphicsContext::drawConsumingImageBuffer(RefPtr<ImageBuffer> image, const
 }
 
-void GraphicsContext::drawFilteredImageBuffer(ImageBuffer* sourceImage, const FloatRect& sourceImageRect, Filter& filter, FilterResults& results)
+void GraphicsContext::drawFilteredImageBuffer(ImageBuffer* sourceImage, const FloatRect& sourceImageRect, Filter& filter, const FilterResultsProvider& resultsProvider)
```
`FilterResultsEnsurer` -> `FilterResultsProvider`

shallawa commented on Source/WebCore/platform/graphics/GraphicsContext.h:
```diff
@@ -273,7 +273,7 @@ class GraphicsContext {
     WEBCORE_EXPORT void drawConsumingImageBuffer(RefPtr<ImageBuffer>, const FloatRect& destination, const FloatRect& source, const ImagePaintingOptions& = { });
 
-    WEBCORE_EXPORT virtual void drawFilteredImageBuffer(ImageBuffer* sourceImage, const FloatRect& sourceImageRect, Filter&, FilterResults&);
+    WEBCORE_EXPORT virtual void drawFilteredImageBuffer(ImageBuffer* sourceImage, const FloatRect& sourceImageRect, Filter&, const FilterResultsProvider& = { });
```
`const FilterResultsProvider& = nullptr` -> `const FilterResultsProvider& = { }`

smfr commented on Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.h:
```diff
@@ -47,10 +47,10 @@ class FilterTargetSwitcher {
 
     virtual void beginClipAndDrawSourceImage(GraphicsContext& destinationContext, const FloatRect&) { beginDrawSourceImage(destinationContext); }
-    virtual void endClipAndDrawSourceImage(GraphicsContext& destinationContext) { endDrawSourceImage(destinationContext); }
+    virtual void endClipAndDrawSourceImage(GraphicsContext& destinationContext, const FilterResultsProvider& resultsProvider = { }) { endDrawSourceImage(destinationContext, resultsProvider); }
```
I'm confused by these default argument functions, which can't return a `FilterResults&`, so what happens if they get called?

shallawa replied:
The `FilterResultsProvider` will be passed down till it reaches `GraphicsContext::drawFilteredImageBuffer()` where `FilterResults` is passed to `Filter::apply()`:
```
    FilterResults results;
    auto result = filter.apply(sourceImage, sourceImageRect, resultsProvider ? resultsProvider() : results);
```
If the default `FilterResultsProvider` is passed this will be equivalent to:
```
    FilterResults results;
    auto result = filter.apply(sourceImage, sourceImageRect, results);
```
This means the `FilterResults` will be discarded once the resulted `ImageBuffer` is generated and drawn to `GraphicsContext`.

shallawa committed:
[GPU Process][Filters] Top level SVGFilter should own its FilterResults
https://bugs.webkit.org/show_bug.cgi?id=256535
rdar://109107572

Reviewed by Simon Fraser.

This allows caching the FilterResults along with the SVGFilter in RemoteResourceCache.

SVGFilter::ensureResults() returns FilterResults& but it takes FilterResultsCreator.
FilterResultsCreator returns a std::unique_ptr<FilterResults> which SVGFilter
will maintain.

* Source/WebCore/platform/graphics/filters/FilterResults.h:
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
(WebCore::RenderSVGResourceFilter::markFilterForRepaint):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.h:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::ensureResults):
(WebCore::SVGFilter::clearEffectResult):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:

Canonical link: https://commits.webkit.org/263996@main
